src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensRangeHandler.cs

```diff
@@ -23,22 +27,80 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler.SemanticTokens
     [ExportRoslynLanguagesLspRequestHandlerProvider(typeof(SemanticTokensRangeHandler)), Shared]
     [Method(Methods.TextDocumentSemanticTokensRangeName)]
-    internal class SemanticTokensRangeHandler : AbstractStatelessRequestHandler<LSP.SemanticTokensRangeParams, LSP.SemanticTokens>
+    internal class SemanticTokensRangeHandler : AbstractStatelessRequestHandler<LSP.SemanticTokensRangeParams, LSP.SemanticTokens>, IDisposable
     {
         private readonly IGlobalOptionService _globalOptions;
-
-        // Lock to guard _projectToCompilation dictionary
-        private readonly object _lock = new();
-        private readonly Dictionary<ProjectId, Task<Compilation?>> _projectIdToCompilation = new();
+        private readonly IAsynchronousOperationListener _asyncListener;
 
         public override bool MutatesSolutionState => false;
         public override bool RequiresLSPSolution => true;
 
+        private readonly object _gate = new();
+        private readonly Dictionary<ProjectId, CompilationAvailableEventSource> _projectIdToEventSource = new();
+        private readonly Dictionary<ProjectId, Checksum> _projectIdLastComputedSolutionChecksum = new();
+
+        // initialized when first request comes in.
+
+        /// <summary>
+        /// Initially null.  Set to true/false when first initialized.  The other related fields will be set if this is true.
+        /// </summary>
+        private bool? _supportsRefresh;
+        private LspWorkspaceManager? _lspWorkspaceManager;
+        private ILanguageServerNotificationManager? _notificationManager;
+        private AsyncBatchingWorkQueue? _semanticTokenRefreshQueue;
+
         [ImportingConstructor]
         [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
-        public SemanticTokensRangeHandler(IGlobalOptionService globalOptions)
+        public SemanticTokensRangeHandler(
+            IGlobalOptionService globalOptions,
+            IAsynchronousOperationListenerProvider asynchronousOperationListenerProvider)
         {
             _globalOptions = globalOptions;
+            _asyncListener = asynchronousOperationListenerProvider.GetListener(FeatureAttribute.Classification);
+        }
+
+        public void Dispose()
+        {
+            ImmutableArray<CompilationAvailableEventSource> eventSources;
+            lock (_gate)
+            {
+                eventSources = _projectIdToEventSource.Values.ToImmutableArray();
+                _projectIdToEventSource.Clear();
+
+                if (_lspWorkspaceManager != null)
+                    _lspWorkspaceManager.LspSolutionChanged -= OnLspSolutionChanged;
+            }
+
+            foreach (var eventSource in eventSources)
+                eventSource.Dispose();
+        }
+
+        [MemberNotNull(nameof(_supportsRefresh))]
+        private void InitializeIfFirstRequest(RequestContext context)
+        {
+            lock (_gate)
+            {
+                if (_supportsRefresh == null)
```

Comment:
I apologize for this mess :(  
Evidence for https://github.com/dotnet/roslyn/issues/61002